### PR TITLE
Create the internal user in a migration

### PR DESCRIPTION
**Pull request by calherries**
The goal of this change is to allow for sample content to be created with the internal user in https://github.com/metabase/metabase/pull/40753, without needing to maintain two ways to create the internal user. 

I decided to create the internal user in all instances, to reduce the number of states the app DB can be in. From now on, all instances will have an internal user.

We create an internal user in a database "migration". The internal user may already exist because it is created in EE instances for the "Metabase Analytics" collection, so we skip creating one if it does.

### CODE REVIEW

File: src/metabase/db/custom_migrations.clj [hunk 1]
```diff
@@ -1068,6 +1070,49 @@
...
+                    :login_attributes nil
+                    :reset_token      nil
+                    :last_name        "Internal"
+                    :date_joined      :%now
+                    :sso_source       nil
+                    :is_datasetnewb   true}]
+      (t2/query {:insert-into :core_user :values [user]}))
+    (let [all-users-id (first (vals (t2/query-one {:select [:id] :from :permissions_group :where [:= :name "All Users"]})))
+          perms-group  {:user_id config/internal-mb-user-id :group_id all-users-id}]
+      (t2/query {:insert-into :permissions_group_membership :values [perms-group]}))))
```
Comment (calherries) [inline src/metabase/db/custom_migrations.clj:1114]: The existing internal user for EE purposes is in the all users perms group, so I do it here too.

File: test/metabase/db/internal_user_test.clj [hunk 2]
```diff
@@ -0,0 +1,41 @@
+(ns ^:mb/once metabase.db.internal-user-test
+  (:require
+   [clojure.string :as str]
+   [clojure.test :refer :all]
+   [metabase.config :as config]
+   [metabase.test :as mt]
+   [metabase.util :as u]
+   [toucan2.core :as t2]))
+
+(deftest internal-user-is-in-all-users-test
```
Comment (calherries) [inline test/metabase/db/internal_user_test.clj:10]: This is a new test

File: test/metabase/db/internal_user_test.clj [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  (testing "The internal user should be in the All Users group, just like any other user."
+    (let [memberships  (-> (t2/select-one :model/User config/internal-mb-user-id)
+                           (t2/hydrate :user_group_memberships)
+                           :user_group_memberships
+                           (->> (map :id)))
+          all-users-id (t2/select-one-pk :model/PermissionsGroup :name "All Users")]
+      (is [all-users-id] memberships))))
+
+(deftest internal-user-is-unmodifiable-via-api-test
```
Comment (calherries) [inline test/metabase/db/internal_user_test.clj:19]: This was moved from enterprise/backend/test/metabase_enterprise/internal_user_test.clj

File: enterprise/backend/src/metabase_enterprise/audit_db.clj [hunk 4]
```diff
@@ -251,7 +250,6 @@
 (defn- maybe-load-analytics-content!
   [audit-db]
   (when analytics-dir-resource
-    (ee.internal-user/ensure-internal-user-exists!)
```
Comment (calherries) [inline enterprise/backend/src/metabase_enterprise/audit_db.clj:254]: No more

File: src/metabase/cmd/copy.clj [hunk 5]
```diff
@@ -199,7 +199,8 @@
   "Make sure [target] application DB is empty before we start copying data."
   [data-source]
   ;; check that there are no Users yet
-  (let [[{:keys [cnt]}] (jdbc/query {:datasource data-source} "SELECT count(*) AS cnt FROM core_user;")]
+  (let [[{:keys [cnt]}] (jdbc/query {:datasource data-source} ["SELECT count(*) AS cnt FROM core_user where not id = ?;"
```
Comment (calherries) [inline src/metabase/cmd/copy.clj:202]: See how this function is used to understand why this is a legal change

File: test/metabase/setup_test.clj [hunk 6]
```diff
@@ -1,23 +1,21 @@
...
-  (testing "The has-user-setup getter should return falsey for an empty instance with only an internal user"
-    ;; create a new completely empty database.
-    (schema-migrations-test.impl/with-temp-empty-app-db [_conn :h2]
-      ;; make sure the DB is set up
-      (mdb/setup-db!)
-      ;; install audit DB, which creates an internal user as a side effect (on EE instances))
-      (mbc/ensure-audit-db-installed!)
-      (is (= false
-             (setup/has-user-setup))))))
+(deftest has-user-setup-ignores-internal-user-test
```
Comment (calherries) [inline test/metabase/setup_test.clj:10]: Effectively replaces `has-user-setup-test` but is now clearer

File: test/metabase/db/internal_user_test.clj [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        (is (= total (count data))))))
+  (testing "User Endpoints with :id"
+    (doseq [[method endpoint status-code] [[:get "user/:id" 400]
+                                           [:put "user/:id" 400]
+                                           [:put "user/:id/reactivate" 400]
+                                           [:delete "user/:id" 400]
+                                           [:put "user/:id/modal/qbnewb" 400]
+                                           [:post "user/:id/send_invite" 400]]]
+      (let [endpoint (str/replace endpoint #":id" (str config/internal-mb-user-id))
+            testing-details-string (str/join " " [(u/upper-case-en (name :get))
```
Comment (crisptrutski) [inline test/metabase/db/internal_user_test.clj:36]: This doesn't look like it should be hardcoded:

```suggestion
            testing-details-string (str/join " " [(u/upper-case-en (name method))
```

If it really is constant, might as well replace with just "GET".

File: test/metabase/db/schema_migrations_test.clj [hunk 8]
```diff
@@ -1180,6 +1181,47 @@
...
+                  :login_attributes nil
+                  :sso_source       nil
+                  :type             "internal"
+                  :date_joined      some?}]
+                (get-users))))))
+  (testing "The internal user isn't created again if it already exists"
+    (impl/test-migrations "v50.2024-03-28T16:30:35" [migrate!]
+      (t2/insert-returning-pks!
+       :core_user
+                               ;; Copied from the old `metabase-enterprise.internal-user` namespace
```
Comment (crisptrutski) [inline test/metabase/db/schema_migrations_test.clj:1207]: This seems strangely indented.

File: test/metabase/db/schema_migrations_test.clj [hunk 9]
```diff
@@ -1180,6 +1181,47 @@
                                         (t2/table-name :model/DataPermissions)
                                         :db_id db-id :group_id group-id :perm_type "perms/manage-database"))))))))
 
+(deftest create-internal-user-test
+  (testing "The internal user is created again if it doesn't already exist"
```
Comment (crisptrutski) [inline test/metabase/db/schema_migrations_test.clj:1185]:
```suggestion
  (testing "The internal user is created if it doesn't already exist"
```

**Review by crisptrutski — APPROVED** [review]
Nice 😎

**Comment by calherries** [comment]
@crisptrutski Thanks for your contributions, I applied them all in c23c2ccb40365c997f55222cfcf398215f806818

**Pull request merged**